### DISCUSSION TITLE
Memory leak for pickle.loads in API endpoint

### DISCUSSION THREAD

**Discussion by korabs-x** [fastapi/fastapi#10879] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import pickle
import uvicorn
from fastapi import FastAPI
import gc

app = FastAPI()

data = pickle.dumps([1] * 50000000)


@app.get("/load-data")
async def load_data():
    pickle.loads(data)
    gc.collect()
    return {"ok?": "ok"}


if __name__ == "__main__":
    uvicorn.run("app.main:app", host="0.0.0.0", port=5000, reload=False)
```

### Description

* Check the memory of the python process (for me ~200MB)
* Open the browser and call http://localhost:5000/load-data several times in quick succession
* Check the memory of the python process (for me ~240MB after making 10 requests in quick succession)
* Repeat the above steps, and watch the memory keep increasing

Expected behaviour:
* In the best case, it should go back to ~200MB
* In any case, it should not accumulate more and more memory when continuing the process
* A way to clean up whatever's leaking here would also help

### Operating System

macOS

### Operating System Details

I can reproduce it locally on Mac, and I can see the same behaviour on a docker image built starting from `python:3.12-slim-bookworm` using GitHub Actions.

### FastAPI Version

0.108.0

### Pydantic Version

2.5.3

### Python Version

Python 3.12.0

**Comment by jgould22** [comment]
I am not too experienced with memory allocators in macOS but I would be quite surprised if the allocator released the memory immediately after a request.

Usually it would expect you to need that memory again and hold on to it.

10 requests is a very small amount of requests for a long running process like a webserver.

What happens if you run a test for 5 minutes constantly making requests?

**Comment by korabs-x** [comment]
Thanks for the reply, I will try to get a graph from mprof showing the change over time.

**Comment by korabs-x** [comment]
I'm sorry, it seems like this code snippet is not enough to reproduce the leak.
I will investigate the leak further and find the smallest reproducable code.
